### PR TITLE
ENH: move exp, log, frexp, ldexp to SIMD dispatching

### PR DESCRIPTION
**Pull request by seiko2plus**
The second patch in a series of pull-requests aims to facilitate the migration
  process to our new SIMD interface(NPYV).

  It is basically a process that focuses on getting rid of the main umath SIMD source `simd.inc`,
  which contains almost all SIMD kernels, by splitting it into several dispatch-able sources without
  changing the base code, which facilitates the review process during the move to NPYV(universal intrinsics).

  In this patch, we have moved the the following raw SIMD loops to the new dispatcher:
   - FLOAT_exp,   DOUBLE_exp
   - FLOAT_log,   DOUBLE_log
   - FLOAT_frexp, DOUBLE_frexp
   - FLOAT_ldexp, DOUBLE_ldexp

### CODE REVIEW

**Comment by mattip** [comment]
Could you benchmark the loops to make sure we did not negatively impact x86_64 performance and if possible on ARM64 as well? 

It seems coverage thinks some of the refactored code is not covered. Is there a test we could add to hit that code path?

File: numpy/core/src/umath/loops_exponent_log.dispatch.c.src [hunk 1]
```diff
@@ -0,0 +1,1293 @@
...
+    return _mm256_mask_i32gather_ps(src, addr, vindex, mask, 4);
+}
+
+static NPY_INLINE __m256d
+fma_masked_gather_pd(__m256d src,
+                     npy_double* addr,
+                     __m128i vindex,
+                     __m256d mask)
+{
+    return _mm256_mask_i32gather_pd(src, addr, vindex, mask, 8);
```
Comment (Qiyu8) [inline numpy/core/src/umath/loops_exponent_log.dispatch.c.src:92]: I'm considering reconstruct `exp/log` by using universal intrinsics so the performance improvement benefits other architecture, but some intrinsics such as `_mm256_mask_i32gather_pd`, `_mm256_blendv_ps`,`_mm512_getmant_ps` is hard to simulate or have a risk to get a slower performance than using scalar.

Comment (mattip) [reply numpy/core/src/umath/loops_exponent_log.dispatch.c.src:92]: @Qiyu8 is this an actionable review comment on this PR? It seems like more of a general statement of intent, correct?

**Comment by mattip** [comment]
LGTM. This is a reshuffle to enable changes in the future, with no real code changes.

**Comment by seiko2plus** [comment]
@mattip,

> Could you benchmark the loops to make sure we did not negatively impact x86_64 performance and if possible on ARM64 as well?

same as before nothing changes

> It seems coverage thinks some of the refactored code is not covered. Is there a test we could add to hit that code path?

all cases are covered by tests but the issue here the coverage runs only for the highest supported SIMD target by the local machine and leaves the other paths untested.

Currently, I'm working on a lightweight execution tracer similar to opencv/#7101, it will allow us to test the dispatcher paths
and also provides elapsed CPU ticks for each execution region.

File: numpy/core/src/umath/loops_exponent_log.dispatch.c.src (hunk 1 above)
Comment (seiko2plus) [reply numpy/core/src/umath/loops_exponent_log.dispatch.c.src:92]: > but some intrinsics such as _mm256_mask_i32gather_pd, _mm256_blendv_ps,_mm512_getmant_ps is hard to simulate or have a risk to get a slower performance than using scalar.

what do you mean hard to simulate? we already did it except for `_mm512_getmant_ps`

`_mm256_mask_i32gather_*` ->  `npyv_loadn_*`  and `npyv_loadn_till_*`
`_mm256_mask_i32scatter_*` -> `npyv_storen_*` and `npyv_storen_till_*`
`_mm256_blend*` -> `npyv_store_*`

> have a risk to get a slower performance than using scalar.

native x86 gather and scatter operations are too expensive and the same for the emulated ones, we should use them without specializing only with large kernels. without specializing -> [loops_trigonometric.dispatch.c.src](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/loops_trigonometric.dispatch.c.src),  with specializing [loops_unary_fp.dispatch.c.src](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/loops_unary_fp.dispatch.c.src)

Comment (seiko2plus) [reply numpy/core/src/umath/loops_exponent_log.dispatch.c.src:92]: note: `_mm256_blend*` blend operations are "not" expensive almost for all SIMD extensions.

**Pull request merged**